first-party/jib-quarkus-extension-gradle/src/main/java/com/google/cloud/tools/jib/gradle/extension/quarkus/JibQuarkusExtension.java

```diff
@@ -0,0 +1,178 @@
+import com.google.cloud.tools.jib.gradle.extension.JibGradlePluginExtension;
+import com.google.cloud.tools.jib.plugins.extension.ExtensionLogger;
+import com.google.cloud.tools.jib.plugins.extension.ExtensionLogger.LogLevel;
+import com.google.cloud.tools.jib.plugins.extension.JibPluginExtensionException;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Verify;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.ResolvedArtifact;
+import org.gradle.api.artifacts.component.ProjectComponentIdentifier;
+import org.gradle.api.plugins.JavaPlugin;
+import org.gradle.jvm.tasks.Jar;
+
+public class JibQuarkusExtension implements JibGradlePluginExtension<Void> {
+
+  private AbsoluteUnixPath appRoot = AbsoluteUnixPath.get("/app");
+  private List<String> jvmFlags = Collections.emptyList();
+
+  @Override
+  public Optional<Class<Void>> getExtraConfigType() {
+    return Optional.empty();
+  }
+
+  @Override
+  public ContainerBuildPlan extendContainerBuildPlan(
+      ContainerBuildPlan buildPlan,
+      Map<String, String> properties,
+      Optional<Void> config,
+      GradleData gradleData,
+      ExtensionLogger logger)
+      throws JibPluginExtensionException {
+    try {
+      logger.log(LogLevel.LIFECYCLE, "Running Quarkus Jib extension");
+      readJibConfigurations(gradleData.getProject());
+
+      Project project = gradleData.getProject();
+      Path buildDir = project.getBuildDir().toPath();
+      Jar jarTask = (Jar) project.getTasks().findByName("jar");
+      String jarName = jarTask.getArchiveFile().get().getAsFile().getName();
+      Path jar = buildDir.resolve(jarName.replaceAll("\\.jar$", "-runner.jar"));
+
+      if (!Files.isRegularFile(jar)) {
+        throw new JibPluginExtensionException(
+            getClass(),
+            jar + " doesn't exist; did you run the Qaurkus Gradle plugin ('quarkusBuild' task)?");
+      }
+
+      ContainerBuildPlan.Builder planBuilder = buildPlan.toBuilder();
+      planBuilder.setLayers(Collections.emptyList());
+
+      // dependency layers
+      addDependencyLayers(project, planBuilder, buildDir.resolve("lib"));
+
+      // Quarkus runner JAR layer
+      AbsoluteUnixPath appRootJar = appRoot.resolve("app.jar");
+      FileEntriesLayer jarLayer =
+          FileEntriesLayer.builder().setName("quarkus jar").addEntry(jar, appRootJar).build();
+      planBuilder.addLayer(jarLayer);
+
+      // Preserve extra directories layers.
+      buildPlan
+          .getLayers()
+          .stream()
+          .filter(layer -> layer.getName().startsWith("extra files"))
```

Comment:
can we get this name from somewhere? or is "extra files" not an available final string anywhere?

Reply:
Good point. The price to pay is to add jib-core as `compileOnly` and that it may go out of sync at Jib runtime (as jib-maven-plugin copies jib-core as source), but I think it's worth paying.